### DISCUSSION TITLE
Questions about AFC implementation for SX127X

### DISCUSSION THREAD

**Discussion by brad112358** [ExpressLRS/ExpressLRS#1996] [General]
Full disclosure, I hope it's OK to ask this here because, in the best open source tradition, ELRS has become an excellent source of example code for other LoRa based projects to reference, including the mLRS project which I use and where I occasionally contribute.

@AlessandroAU,
I recently decided to implement AFC for mLRS and I noticed how different the ELRS implementation in HandleFreqCorr() is from the method of obtaining the frequency correction documented in section 4.1.6 of the SX1276 datasheet.  You seem to have implemented this early on and I can't find any indication in the ELRS code or git history that direct use of the entire the LoRaFeiValue, as described in the datasheet was ever attempted before this iterative method which uses only the sign bit of the RegFeiMsb register to move the frequency slowly over time was implemented.

Was the direct method described in the datasheet ever tested for ELRS when it was being developed?  Was there a problem observed or predicted with this method?
Was the iterative method implemented in HandleFreqCorr() recommended by an Semtech application note or application engineer?

I do like the iterative method as it requires fewer register reads and I think it would tend to minimize problems from any spurious error readings as it would smooth out any big jumps.  It might even provide more accurate results by the measurement averaging effect it has.  I have implemented a variant of this iterative method for mLRS and it does seem to be working well there but I'd love to know the genesis of this approach.

Thanks!

**Comment by JyeSmith** [comment]
Ill also add my 2c since Ive been poking around the sx127x recently.

2 points that you've already made stand out to me...

- Minimising SPI traffic is beneficial in maximising the time available for packet tx/rx.  Then we can use better modulation params for a stronger link.

- Smoothing changes. Once we have a connection the changes in offset arent huge.  Below is a log of the incremental change from power up.  Once its settled we know there wont be large changes and the offset just bounces around.

![image](https://user-images.githubusercontent.com/14170229/207965306-13994649-d9fc-4ed2-b055-95dec90f8275.png)

Even with heat the change is minimal when compared to the original offset on power up. Red is another log with hotair set to 100C cooking the rx.

![image](https://user-images.githubusercontent.com/14170229/207966533-0deaaac7-2d16-47bc-9ea8-6f878dc65edb.png)

**Comment by brad112358** [comment]
Thanks for the response Jye!
The graphs you show are exactly what I expected with this method.
I am still curious though if the direct single step, full error estimate correction method specified in the datasheet was tried early on and if it resulted in a more noisy/jittery error estimate such that the current method improved on more than just the spi traffic.  Perhaps only @AlessandroAU knows the earliest history of this.

**Reply by JyeSmith** [reply]
Thats something left for the reader to test 😄 

I imagine even if the jitter is less it will mean SFA to final reception of a packet.

**Comment by AlessandroAU** [comment]
You guys covered pretty much all the important bits already :) but yeah with the full correction on each frame it is not stable because sometimes the freq error of a single packet is not returned correctly. You could make an argument to do full correction immediately then switch to the incremental mode, but the time it takes to lock on is so short anyway I don't think it's worth the added complexity.

**Comment by brad112358** [comment]
Thanks @AlessandroAU!  One mystery explained.  :)

**Comment by brad112358** [comment]
Since we are discussing this anyway;  When I was reviewing the ELRS implementation of AFC, I noticed something that might not be optimal and a couple of things that I don't yet see a good reason for.  It appears that you only change the frequency and take the error estimate into account when it's time for the frequency to hop.  This the error estimate will accumulate a few increments/decrements based on the same frequency setting and correction, depending on the hop interval.  I think this might be what is causing the small fairly regular oscillation visible in the charts you included above.  This may not cause much problem in practice as the overshoot should remain less than the hop interval.  It also looks like you only increment/decrement error estimate when you don't hop and I'm not sure why.  If you don't want to correct the frequency when you don't need to write the register anyway, would it be better find a way to average of the error sign samples taken since the last hop?
I'm also not sure why you reset the error estimate on connection loss.

**Comment by JyeSmith** [comment]
When you plot out and compare the freq offset for the current method and the correct way there is no difference.  They are both really noisy.

Doing it the correct way and waiting for the hop also means the settle time is much longer.

**Reply by brad112358** [reply]
OK.  Thanks!
